Clean up SCTP tests

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Cleans up some ugliness in the SCTP e2e tests.

#### Which issue(s) this PR fixes:
Fixes #96717

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/priority backlog
/sig network

danwinship committed:
Replace iptables-proxy-specific SCTP e2e test with a unit test

We had a test that creating a Service with an SCTP port would create
an iptables rule with "-p sctp" in it, which let us test that
kube-proxy was doing vaguely the right thing with SCTP even if the e2e
environment didn't have SCTP support. But this would really make much
more sense as a unit test.

danwinship committed:
Drop iptables-specific test from SCTP HostPort test

The SCTP HostPort test was checking that creating a pod with an SCTP
HostPort would create a certain iptables rule, but the handling of
HostPorts is now up to CRI, not kubelet, so kubernetes e2e cannot
assume it will implement the feature in any specific way.

(The test still ensures that (a) the apiserver accepts SCTP HostPorts,
and (b) neither kubelet nor the runtime causes the SCTP kernel module
to be loaded as part of creating a pod with an SCTP HostPort.)

danwinship committed:
Remove [Disruptive] from Feature:SCTPConnectivity tests

The Feature:SCTPConnectivity tests cannot run at the same time as the
"X doesn't cause sctp.ko to be loaded" tests, since they may cause
sctp.ko to be loaded. We had dealt with this in the past by marking
them [Disruptive], but this isn't really fair; the problem is more
with the sctp.ko-checking tests than it is with the SCTPConnectivity
tests. So make them not [Disruptive] and instead make the
sctp.ko-checking tests be [Serial].

aojea commented:
+1 on the other commits but `Consolidate SCTP API / non-module-loading tests` , I think that the classification is based on the hostPort/NetworkPolicy/Service and SCTP is just a variation. If any of those fails, the whole test will fail and the user will have to check the logs to understand what happens.

danwinship commented:
pull-kubernetes-e2e-ubuntu-gce-network-policies is failing on a newly-non-`[Disruptive]` `[Feature:SCTPConnectivity]` test. I haven't dug into it, but the obvious guess would be that the e2e job is running the `SCTPConnectivity` tests despite not having SCTP connectivity

danwinship commented:
https://github.com/kubernetes/test-infra/pull/27858 will fix the network-policy job to explicitly skip the SCTPConnectivity tests (which don't work in that environment because Calico doesn't support SCTP)

aojea commented on test/e2e/network/networking.go:
```diff
@@ -646,28 +646,35 @@ var _ = common.SIGDescribe("SCTP [LinuxOnly]", func() {
+		sctpLoadedAtStart = CheckSCTPModuleLoadedOnNodes(f, nodes)
+	})
+
+	ginkgo.AfterEach(func() {
```
you may prefer to use `DeferCleanup` for this?

https://onsi.github.io/ginkgo/#cleaning-up-our-cleanup-code-defercleanup

danwinship replied:
I ended up removing the BeforeEach-refactoring commit; if we're not moving the NetworkPolicy test to be with the other two, then it makes less sense to keep them together as well; so I just left the SCTP service test in service.go and moved the SCTP hostport tests to networking.go (and left the SCTP NetworkPolicy test in netpol), and now there's no shared code to refactor out.

aojea commented:
/retest
/approve

aojea commented:
/test

aojea commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd-serial

aojea commented:
/retest
🤔

aojea commented:
/hold

this PR is making the pull-kubernetes-e2e-ubuntu-gce-network-policies	 job to timeout,

it is passing in another PRs per the history https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-ubuntu-gce-network-policies

danwinship commented:
looking at the top of the build-log, it's still using

`-ginkgo.skip=\[Feature:(Networking-IPv6|Example|Federation|PerformanceDNS|KubeProxyDaemonSetMigration|ProxyTerminatingEndpoints)\]|DualStack|GCE|Disruptive|Serial|SNAT|LoadBalancer|ESIPP'`

without `SCTPConnectivity`. So my test-infra PR hasn't fully propagated yet.

aojea commented:
@danwinship my bad, it was removed only from the periodic job https://github.com/kubernetes/test-infra/commit/5986e87906906cab8dd32df8f35c204407d618ee

you have to remove it from the presubmit job too, it should be up on the same file, tag me and I´ll approve quickly

danwinship committed:
Move two SCTP tests

There were two SCTP tests grouped together in
test/e2e/network/service.go, but one of them wasn't a service test...
so move the SCTP service test to be grouped with the other service
tests, and the SCTP hostport tests to be grouped with other
non-service tests.

danwinship commented:
/test all

dims commented:
If you still need this PR then please rebase, if not, please close the PR

aojea commented:
/lgtm
/hold cancel



ipvs job failures https://github.com/kubernetes/kubernetes/issues/114450
network policies failures https://github.com/kubernetes/kubernetes/issues/114449
